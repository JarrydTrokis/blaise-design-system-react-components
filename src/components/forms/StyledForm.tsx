import React, {Fragment} from 'react';
import {Field, Form, Formik} from "formik";
import ErrorSummary from './ErrorSummary';
import {ONSInputField, InputErrorPanel} from "./InputField";


export interface FormFieldObject {
    name: string
    description?: string
    type: string
    validate?: (value: string) => string | undefined
    autoFocus?: boolean
}

export interface StyledFormProps {
    fields: FormFieldObject[]
    onSubmitFunction: (values: any, setSubmitting: (isSubmitting: boolean) => void) => void
}

/**
 * Formik form styled to ONS design guide with form error panel
 *
 * @param Props
 *
 *  - fields: List of fields to display on form.
 *  - onSubmitFunction: Function to call after submit of form and all field validation is valid.
 */
function StyledForm({fields, onSubmitFunction}: StyledFormProps) {


    let initialFieldValues: any = {};
    fields.forEach(({name}) => {
        initialFieldValues[name] = "";
    });

    return (
        <Formik
            validateOnBlur={false}
            validateOnChange={false}
            initialValues={initialFieldValues}
            onSubmit={(values, {setSubmitting}) => {
                onSubmitFunction(values, setSubmitting);
            }}
        >
            {({
                  errors,
                  isSubmitting,
                  isValid
              }) => (
                <Form>
                    {ErrorSummary(isValid, errors)}

                    {
                        fields.map((field, index) => {
                            field.autoFocus = (index === 0)
                            return (
                                <Fragment key={field.name}>
                                    {
                                        // @ts-ignore
                                        errors[field.name] ?
                                            // @ts-ignore
                                            InputErrorPanel(
                                                // @ts-ignore
                                                errors[field.name],
                                                "name",
                                                <Field {...field} component={ONSInputField}/>
                                            )
                                            :
                                            <Field {...field} component={ONSInputField}/>
                                    }
                                </Fragment>
                            )
                        })
                    }

                    <br/>

                    <button type="submit" className="btn" disabled={isSubmitting} data-testid={"submit-btn"}>
                        <span className="btn__inner">Save and continue</span>
                    </button>
                </Form>
            )}
        </Formik>
    )
}

export default StyledForm;