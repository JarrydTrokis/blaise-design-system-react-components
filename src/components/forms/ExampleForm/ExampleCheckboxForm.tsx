import React, { useState } from "react";
import StyledForm, { FormFieldObject } from "../StyledForm";
import { validateCheckbox } from "./FormValidation";

function ExampleForm() {
    const [formStatus, setFormStatus] = useState<string>("");

    /** List of fields in order for form generation */
    const formElements: FormFieldObject[] = [
        {
            name: "questionnaire",
            description: "Select questionnaires",
            type: "checkbox",
            validate: validateCheckbox,
            checkboxOptions: [
                { id: "lms", value: "lms", label: "LMS" },
                { id: "opn", value: "opn", label: "OPN" },
            ],
        },
    ];

    /**
     * Function is called after submit of form and all field validation is valid
     *
     * @param formValues Object with all field values
     * @param setSubmitting Function to set isSubmitting attribute which disables submit button while processing the form
     */
    function onFormSubmission(formValues: any, setSubmitting: (isSubmitting: boolean) => void): void {
        console.warn(formValues);
        setFormStatus(`Form submitted, questionnaires chosen ${formValues.questionnaire}`);
        setSubmitting(false);
    }

    return (
        <>
            <p>{formStatus}</p>
            <StyledForm fields={formElements} onSubmitFunction={onFormSubmission} />
        </>
    );
}

export default ExampleForm;
