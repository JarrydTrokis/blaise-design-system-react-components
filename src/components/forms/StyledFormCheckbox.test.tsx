import React from "react";
import {
    fireEvent, render, screen, waitFor,
} from "@testing-library/react";
import ExampleCheckboxForm from "./ExampleForm/ExampleCheckboxForm";
import StyledForm from "./StyledForm";

test("it matches snapshots", async () => {
    const formPage = render(<ExampleCheckboxForm/>);

    expect(formPage).toMatchSnapshot();
});

test("error appears on submit of empty form", async () => {
    render(<ExampleCheckboxForm/>);

    fireEvent.click(screen.getByTestId(/submit-button/i));

    await waitFor(() => {
        const errorMessage = screen.getByText(/There is 1 problem with your answer/i);
        expect(errorMessage).toBeInTheDocument();

        const checkboxErrorMessage = screen.queryAllByText(/Select an option/i);
        expect(checkboxErrorMessage).toHaveLength(2);
    });
});

test("submit function is called when one checkbox is selected", async () => {
    render(<ExampleCheckboxForm/>);

    fireEvent.click(screen.getByLabelText(/LMS/i));

    fireEvent.click(screen.getByTestId(/submit-button/i));

    await waitFor(() => {
        const successMessage = screen.getByText(/Form submitted, questionnaires chosen lms/i);
        expect(successMessage).toBeInTheDocument();
    });
});

test("submit function is called when multiple checkboxes are selected", async () => {
    render(<ExampleCheckboxForm/>);

    fireEvent.click(screen.getByLabelText(/LMS/i));
    fireEvent.click(screen.getByLabelText(/OPN/i));

    fireEvent.click(screen.getByTestId(/submit-button/i));

    await waitFor(() => {
        const successMessage = screen.getByText(/Form submitted, questionnaires chosen lms,opn/i);
        expect(successMessage).toBeInTheDocument();
    });
});

test("setting initial value", async () => {
    const fields = [
        {
            name: "topping",
            description: "Select your favorite topping",
            type: "checkbox",
            initial_value: ["bacon", "pineapple"],
            checkboxOptions: [
                { id: "bacon", value: "bacon", label: "Bacon" },
                { id: "cheese", value: "cheese", label: "Cheese" },
                { id: "pineapple", value: "pineapple", label: "Pineapple" },
            ],
        },
    ];

    const submitFunction = jest.fn();

    render(<StyledForm fields={fields} onSubmitFunction={submitFunction}/>);

    fireEvent.click(screen.getByTestId(/submit-button/i));

    await waitFor(() => {
        expect(submitFunction).toBeCalledWith(
            expect.objectContaining({ "topping": ["bacon", "pineapple"] }),
            expect.any(Function),
        );
    });
});

test("all checkboxes are selected when select all button is clicked", async () => {
    render(<ExampleCheckboxForm/>);

    fireEvent.click(screen.getByText(/Select all/i));

    fireEvent.click(screen.getByTestId(/submit-button/i));

    await waitFor(() => {
        const successMessage = screen.getByText(/Form submitted, questionnaires chosen lms,opn/i);
        expect(successMessage).toBeInTheDocument();
    });
});

test("all checkboxes are deselected when unselect all button is clicked", async () => {
    render(<ExampleCheckboxForm/>);

    fireEvent.click(screen.getByText(/Select all/i));
    fireEvent.click(screen.getByText(/Unselect all/i));

    fireEvent.click(screen.getByTestId(/submit-button/i));

    await waitFor(() => {
        const errorMessage = screen.getByText(/There is 1 problem with your answer/i);
        expect(errorMessage).toBeInTheDocument();

        const checkboxErrorMessage = screen.queryAllByText(/Select an option/i);
        expect(checkboxErrorMessage).toHaveLength(2);
    });
});
