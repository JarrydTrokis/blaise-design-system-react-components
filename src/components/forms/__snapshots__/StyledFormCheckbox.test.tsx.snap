// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it matches snapshots 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <p />
      <form
        action="#"
      >
        <fieldset
          class="fieldset"
        >
          <legend
            class="fieldset__legend"
          >
            Select questionnaires
          </legend>
          <button
            class="btn u-mb-s js-auto-selector btn--small btn--secondary"
            type="button"
          >
            <span
              class="btn__inner"
            >
              <span
                class="js-button-text"
              >
                Select All
              </span>
              <span
                class="u-vh"
              >
                 following checkboxes
              </span>
            </span>
          </button>
          <div
            class="checkboxes__items"
            id="questionnaire"
          >
            <p
              class="checkboxes__items"
            >
              <span
                class="checkbox"
              >
                <input
                  class="checkbox__input js-checkbox"
                  id="lms"
                  name="questionnaire"
                  type="checkbox"
                  value="lms"
                />
                <label
                  class="checkbox__label "
                  for="lms"
                  id="lms-label"
                >
                  LMS
                </label>
              </span>
            </p>
            <br />
            <p
              class="checkboxes__items"
            >
              <span
                class="checkbox"
              >
                <input
                  class="checkbox__input js-checkbox"
                  id="opn"
                  name="questionnaire"
                  type="checkbox"
                  value="opn"
                />
                <label
                  class="checkbox__label "
                  for="opn"
                  id="opn-label"
                >
                  OPN
                </label>
              </span>
            </p>
            <br />
          </div>
        </fieldset>
        <br />
        <button
          class="btn "
          data-testid="submit-button"
          type="submit"
        >
          <span
            class="btn__inner"
          >
            Save and continue
          </span>
        </button>
      </form>
    </div>
  </body>,
  "container": <div>
    <p />
    <form
      action="#"
    >
      <fieldset
        class="fieldset"
      >
        <legend
          class="fieldset__legend"
        >
          Select questionnaires
        </legend>
        <button
          class="btn u-mb-s js-auto-selector btn--small btn--secondary"
          type="button"
        >
          <span
            class="btn__inner"
          >
            <span
              class="js-button-text"
            >
              Select All
            </span>
            <span
              class="u-vh"
            >
               following checkboxes
            </span>
          </span>
        </button>
        <div
          class="checkboxes__items"
          id="questionnaire"
        >
          <p
            class="checkboxes__items"
          >
            <span
              class="checkbox"
            >
              <input
                class="checkbox__input js-checkbox"
                id="lms"
                name="questionnaire"
                type="checkbox"
                value="lms"
              />
              <label
                class="checkbox__label "
                for="lms"
                id="lms-label"
              >
                LMS
              </label>
            </span>
          </p>
          <br />
          <p
            class="checkboxes__items"
          >
            <span
              class="checkbox"
            >
              <input
                class="checkbox__input js-checkbox"
                id="opn"
                name="questionnaire"
                type="checkbox"
                value="opn"
              />
              <label
                class="checkbox__label "
                for="opn"
                id="opn-label"
              >
                OPN
              </label>
            </span>
          </p>
          <br />
        </div>
      </fieldset>
      <br />
      <button
        class="btn "
        data-testid="submit-button"
        type="submit"
      >
        <span
          class="btn__inner"
        >
          Save and continue
        </span>
      </button>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
